src/python/pants/backend/go/util_rules/third_party_pkg.py

```diff
@@ -327,23 +330,77 @@ async def _check_go_sum_has_not_changed(
             )
 
 
+@rule
+async def process_find_go_packages(
+    request: ModulePackageListRequest,
+    bash: BashBinary,
+    find: FindBinary,
+) -> ModulePackageList:
+    """This connects the (cached) go mod download steps with the package analyzer steps by finding
+    likely package directories within a module."""
+
+    # As background, several other options did not work for this layer:
+    # - `go list` provides the same package paths, but fails on modules which do not use go.sum
+    # - `os.walk` could process the named cache folder directly, but running a Process allows caching
+
+    find_script = FileContent(
+        "__run_find.sh",
+        textwrap.dedent(
+            f"""\
+            {find.path} {request.path} -type f -name '*.go'
+            """
+        ).encode("utf-8"),
+    )
+
+    digest = await Get(Digest, CreateDigest([find_script]))
```

Comment:
This can be moved to a separate rule which returns a "setup" struct. (This pattern is used elsewhere in the Pants source.) That will allow Pants to cache the creation of the digest and not have to re-run the digest creation each time the package query occurs.